### PR TITLE
Remove usage of --no-search-upwards in tests/samples

### PR DESCRIPTION
**Pull request by bmuschko**
### Context

See https://github.com/gradle/gradle/issues/3333

- Adds `settings.gradle` files to sample projects to prepare for good practice.
- Introduces system property for Gradle versions going forward to prevent the use of `--no-search-upwards`.
- Requires the use of `--no-search-upwards` for cross-version tests with older versions of Gradle (that do not know about the system property yet) e.g. `WrapperCrossVersionIntegrationTest`. Let me know if there's a better option.
- Removes references to `--no-search-upwards` where possible.
- TestKit already used the internal method `DefaultGradleConnector.searchUpwards`.

### CODE REVIEW

**Comment by bmuschko** [comment]
There's still something wrong with the cross-version tests when I actually deprecate the option. I somewhat doesn't make sense to separate this change from the actual deprecation. We'll need to know if it actually works so I will probably put that into the same PR.

File: subprojects/core/src/integTest/groovy/org/gradle/initialization/CommandLineArgDeprecationIntegrationTest.groovy [hunk 1]
```diff
@@ -25,11 +25,12 @@ class CommandLineArgDeprecationIntegrationTest extends AbstractIntegrationSpec {
     private static final String RECOMPILE_SCRIPTS_MESSAGE = '--recompile-scripts has been deprecated and is scheduled to be removed in Gradle'
 
     private static final String NO_REBUILD_MESSAGE = '--no-rebuild/-a has been deprecated and is scheduled to be removed in Gradle'
+    private static final String NO_SEARCH_UPWARD_MESSAGE = '--no-search-upward/-u has been deprecated and is scheduled to be removed in Gradle'
 
     @Unroll
     def "deprecation warning appears when using #deprecatedArgs"() {
         when:
-        executer.expectDeprecationWarning()
+        executer.expectDeprecationWarning().requireGradleDistribution()
```
Comment (oehme) [inline subprojects/core/src/integTest/groovy/org/gradle/initialization/CommandLineArgDeprecationIntegrationTest.groovy]: ⭕️  Why do we need a distribution for this?

File: subprojects/wrapper/src/crossVersionTest/groovy/org/gradle/integtests/WrapperCrossVersionIntegrationTest.groovy [hunk 2]
```diff
@@ -104,6 +104,7 @@ task hello {
     }
 }
 """
+        settingsFile << ''
```
Comment (oehme) [inline subprojects/wrapper/src/crossVersionTest/groovy/org/gradle/integtests/WrapperCrossVersionIntegrationTest.groovy]: ⭕️ Is this still necessary after you introduced the property? (same for the other occurrences above)

**Comment by bmuschko** [comment]
> If we want to show good practice, the settings.gradle files in the samples should set the root project name. This can be done in a follow-up change though.

Let's do that with a separate change.

File: subprojects/core/src/integTest/groovy/org/gradle/initialization/CommandLineArgDeprecationIntegrationTest.groovy (hunk 1 above)
Comment (bmuschko) [reply subprojects/core/src/integTest/groovy/org/gradle/initialization/CommandLineArgDeprecationIntegrationTest.groovy]: It looks like the deprecation message is emitted before we even capture standard output. That might be reason why we [batch up the deprecation messages in `StartParameterInternal`](https://github.com/gradle/gradle/blob/master/subprojects/core/src/main/java/org/gradle/api/internal/StartParameterInternal.java#L44) in then [check them later](https://github.com/gradle/gradle/blob/master/subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/StartParamsValidatingActionExecuter.java#L65-L67)?

File: subprojects/wrapper/src/crossVersionTest/groovy/org/gradle/integtests/WrapperCrossVersionIntegrationTest.groovy (hunk 2 above)
Comment (bmuschko) [reply subprojects/wrapper/src/crossVersionTest/groovy/org/gradle/integtests/WrapperCrossVersionIntegrationTest.groovy]: I am going to go through them one by one again and will leave the reasoning here:

- `ToolingApiIntegrationTest."tooling API does not hold JVM open"`: We run a build with the `GradleExecutor` which in turn runs a Java main class. That main class uses the Tooling API to execute a task of the same build script (which doesn't have a settings file) and will search upward. The alternative is to use `((DefaultGradleConnector) connector).searchUpwards(false);` here which might make the intention more clear. Changed to that.

File: subprojects/core/src/integTest/groovy/org/gradle/initialization/CommandLineArgDeprecationIntegrationTest.groovy (hunk 1 above)
Comment (oehme) [reply subprojects/core/src/integTest/groovy/org/gradle/initialization/CommandLineArgDeprecationIntegrationTest.groovy]: Logging is set up [before](https://github.com/gradle/gradle/blob/ac48633eab60f328da4fd6a080c31dda43b60b91/subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/LauncherServices.java#L74) that, so it should work. Bo also removed the `requireGradleDistribution` from all the other warning tests, so I wonder why it wouldn't work in this one.

Comment (bmuschko) [reply subprojects/core/src/integTest/groovy/org/gradle/initialization/CommandLineArgDeprecationIntegrationTest.groovy]: Not quite sure. You can even see the deprecation message in the console output if you remove `requireGradleDistribution()`. Any other ideas why that could be?

**Comment by bmuschko** [comment]
@oehme I commented on your feedback. Let's chat about soon.

**Pull request merged**